2.x Fix can_edit() permission checks for Term, User, Comment and Menu classes

gchtr opened the pull request:
Resolves:

- #2645

## Issue

When using `{{ term.can_edit }}`, Timber should use the correct permission check for a term. The same goes for other core objects like comment and users.

## Solution

This pull request

- Moves `can_edit()` to the specific Timber object classes.
- Adds `can_edit()` to the `CoreEntity` interface.
- Adds new `edit_link()` functions to `Comment` and `User` class.

## Impact

Accurate capability checks.

## Usage Changes

None.

## Considerations

Checking the permissions on the `Menu` and `MenuItem` feels a bit weird first. They can’t be edited through the usual edit screens and need a check for the `edit_theme_options` capability. The question now is: Does it make sense to have the `can_edit()` check for the menu and menu items? Otherwise, we would have to remove the interface.

By adding `User::edit_link()`, it’s possible to get the profile URL for the current user, which is nice! I wonder whether we should add a `User::profile_link()` method as well. This would only return a link if the user object is the current user.

## Testing

Yes.

nlemoine reviewed (changes requested):
> Does it make sense to have the can_edit() check for the menu and menu items?

It doesn't from a usage POV but it does from technical. I would stand by the technical POV. `Menu` and `MenuItem` are special terms and posts and should implement the `CoreEntity` interface even if it's a bit weird in that one use case.

> I wonder whether we should add a User::profile_link() method as well

This makes sense!

Besides adding return types to the `edit_link` methods, this looks good to me!

gchtr commented:
> > I wonder whether we should add a User::profile_link() method as well
> 
> This makes sense!

Added as a future task in https://github.com/timber/timber/issues/2682.

> Besides adding return types to the `edit_link` methods, this looks good to me!

Yes, that makes total sense! Added in [`db206bb` (#2676)](https://github.com/timber/timber/pull/2676/commits/db206bb31cbaac024709f51f366d39381852a11c).

This is now ready for a final review.

nlemoine reviewed (approved):
All good to me! Thanks @gchtr

gchtr commented:
@nlemoine Thanks for the review! I’ll merge this in.
